### ISSUE TITLE
macOS: SecKeyCreateSignature fails for digest input with "SmartCard returned error 6982"

### ISSUE THREAD

**Issue by gnida-rada** [Yubico/yubico-piv-tool#302]
Hello,
We have a tooling to install a certificate into slot 9a. The private key was generated with RSA2048 algorithm.
Later we use that cert to sign some data. I am using Apple's SecKeyCreateSignature with one of the two algorithms:
kSecKeyAlgorithmRSASignatureMessagePKCS1v15SHA256 (that first creates digest then signs passed plain data) and
kSecKeyAlgorithmRSASignatureDigestPKCS1v15SHA256 (the caller is expected to create the digest, so Apple's API signs the digest)
When I call SecKeyIsAlgorithmSupported(privateKeyfromYubiKey, <algorithm>, kSecKeyOperationTypeSign), I always get TRUE back.
When I call SecKeyCreateSignature with  kSecKeyAlgorithmRSASignatureMessagePKCS1v15SHA256, it correctly creates signature.
However, when I call SecKeyCreateSignature with kSecKeyAlgorithmRSASignatureDigestPKCS1v15SHA256, it returns nil and reports "SmartCard returned error 6982".

I did some additional experiments: created a self-signed cert on either a different slot (9c, 9e) or on the same slot 9a.
Interestingly: at first, I was prompted by macOS for PIN several times and signing with either algorithm succeeded and the signatures from either approach were identical (as they should).
But on re-tries the same symptoms repro: signing plain data works, signing digest - fails with "SmartCard returned error 6982". SecKeyIsAlgorithmSupported always, always returns TRUE for either algorithm!

Number 6982 does not give me much to work with. 
- Am I passing something wrong? Unlikely. the same code works fine with certs on another (non-Yubico) Smart Card.
- Is Apple passing something wrong to YubiKey? Again, the same code works with certs from Keychain and from another Smart Card.
- Is there a bug in YubiKey code?
- Is there a better place for me to discuss this problem?
I'm happy to supply a tiny test app if it helps to diagnose the issue. 
I work with YubiKey FIPS.
Thanks!

**Comment by a-dma** [comment]
Hi,

those are ISO 7816 SW. Specifically 0x6982 is the error code SECURITY_STATUS_NOT_SATISFIED which means that the necessary authentication hasn't happened. There are two things to keep in mind here one is that different operations require different types of authentication (Management key vs PIN), see [here](https://developers.yubico.com/PIV/Introduction/Admin_access.html), the other is that different slots have different _default_ PIN policies see [here](https://developers.yubico.com/PIV/Introduction/Certificate_slots.html). Default PIN policies can be changed, but only during key generation.

That said... what does your Apple API use to talk to the device? Does it send APDU directly to the device? Does it use a library like libykpiv? Does it use PKCS#11? Try to get some debug output of which commands are actually sent to the device at the end of the whole chain.

Last but not least, there is no direct support for PKCS#1 v1.5 (or other padding schemes) in the PIV application itself. It's up to the user to hash and pad the data correctly before it is sent to the device. Chances are that some incorrect data is sent out but that a misleading error is surfaced. Hard to tell without being able to see the APDUs sent to the YubiKey.

**Comment by gnida-rada** [comment]
Thanks for your reply, @a-dma, it helped to move forward with investigation!
**Update**: it looks like it has nothing to do with specific algorithm, but rather with 
a) # of times a process attempts to codesign using a private key from Yubikey and 
b) default PIN policy.
Specifically, if I generate key and certificate on any slot other than 9c, I have issues with code signing even when I set --pin-policy=always
SLOT=9a
$YKMAN piv generate-key -a RSA2048 --touch-policy=never **--pin-policy=always** -P $DEFAULT_PIN -m $DEFAULT_MK $SLOT $PUBLIC_KEY_PATH
$YKMAN  piv generate-certificate -s "CN=yubikey_test,OU=Yubikey Test,DC=one,DC=two,DC=three,DC=com" -P $DEFAULT_PIN -m $DEFAULT_MK $SLOT $PUBLIC_KEY_PATH

If I try to codesign **twice** using private key from Yubikey:
- if the key/cert are on slot **9c**, I get prompted for PIN 2 times and both times code signing succeeds.
- if the key/cert are on slot **9a**, I get prompted for PIN (which looks like default policy for 9a) only once and then 
- - first code signing succeeds.
- - second code signing fails with "SmartCard returned error 6982" 
- for similar key/cert on slot **9d**, Apple's SecKeyIsAlgorithmSupported  returns FALSE
- for similar key/cert on slot **9e**, I never get prompted for PIN and both attempts to code sign result in "SmartCard returned error 6982"

So it looks like that, despite me passing -pin-policy=always to
ykman piv generate-key
it follows the default PIN policies. Is that by design or a bug on Yubikey side?

**Comment by a-dma** [comment]
I suspect that the backend is aware of the default PIN policies and it's trying to be too helpful. Out of curiosity, what happens if you generate keys with a pin policy set to `never`?

Alternatively you can try signing data directly with something you control entirely (like using libykpiv directly for example) and see if that behaves in the expected way.

**Comment by mouse07410** [comment]
Bring unable to help with debugging this, I'd like to suggest a workaround. 

In general, 9a is a shot for the Authentication key, and 9c - for Signature key. That means - 9c is the one that should be used for codesigning. Would you consider just using 9c?

@a-dma  the two PIN prompts are for the token itself and the key slot. In most cases it dies not make sense, as the PIN value is the same, and there's no operation on the token between the token login and signing. Is there a way to avoid this double prompt?

**Comment by a-dma** [comment]
I'm not sure what you mean with "the token itself", but generally speaking the device is never prompting for a PIN. It has an internal state and upon receiving a command that asks it to sign some data it checks whether it is allowed to do so or not, and responds accordingly. If the application has some internal logic to determine that it should prompt the user for a PIN twice I'm afraid there isn't much the device can help with.

That's why a think that getting a trace of what is being sent to the device can help, perhaps the application prompts the user as a reaction to an error, or perhaps it does it based on which slot is used, or something similar.

**Comment by mouse07410** [comment]
> I'm not sure what you mean with "the token itself", but generally speaking the device is never prompting for a PIN.

Middleware usually requires login to the "token". That's what I call when, e.g., you need to authenticate using key in 9a, and "it" (Firefox, OpenSC, whatever) prompts for you a PIN. Then, there are slots with "PIN policy = always" - then "it" will prompt you *again* to enter your PIN, presumably for the *key slot* (as it is *possible* to have a separate PIN for such a key/slot). 

My OpenSC fork avoids doing this, and if the card considers the state "Logged in", the middleware won't bother the user.

**Comment by gnida-rada** [comment]
@a-dma , you mentioned PIN policies described [here](https://developers.yubico.com/PIV/Introduction/Certificate_slots.html). That page is saying "A PIV-enabled YubiKey NEO holds 4 distinct slots for certificates and a YubiKey 4 & 5 holds 24, as specified in the [PIV standards document](https://csrc.nist.gov/Projects/PIV/PIV-Standards-and-Supporting-Documentation). "
That "PIV standards document" is a repository of documents. I downloaded 10 or so .pdfs and none of them was talking about certificate slots. Do you know what document is covering cert slots?

Thanks for the suggestions. I'll play with 
- lower level code signing APIs
- token-policy set to never and report the findings here.

**Comment by a-dma** [comment]
That would be [SP 800 73-4](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-73-4.pdf)

If you're interested in using `libkykpiv` from this repository, have a look at https://github.com/Yubico/yubico-piv-tool/blob/master/lib/tests/api.c or at the source. There's a lot of boilerplate since those are tests rather than examples, but it should get you started. The source of `yubico-piv-tool` itself is also a possibility.

**Comment by gnida-rada** [comment]
So I guess I need to clarify something first: the overriding of PIN-policy is not a standard behavior, right?
And if so, if developers want to take advantage on it, they should be using yubiko library and not (most likely standard compliant) system library. 
@a-dma ,  is that correct? Thanks

**Comment by mouse07410** [comment]
As I explained, IMHO developers should use the key in 9c for code signing, not in 9a.

**Comment by a-dma** [comment]
@gnida-rada yes, this is the list of additions that are not part of the standard https://developers.yubico.com/PIV/Introduction/Yubico_extensions.html.

As to whether it is OK to use a library that is shipped with the system, I don't really have a better answer other than to try it. It might be that some additions were picked up by other tools. Yubico's library support the ones linked above.

**Comment by qpernil** [comment]
To give a bit of update here - the latest release does indeed read the PIN policies from the YubiKey so now you should get proper values for CKA_ALWAYS_AUTHENTICATE. This has som limitations however - if keys are imported a YubiKey version >= 5.3.0 is required to be able to read this info. For generated keys it is enough to have a YubiKey 4+.
There are also proprietary attributes that can be read (via ykcs11), see pkcs11y.h, to get the full policy info (CKA_ALWAYS:AUTHENTICATE only tells part of the story).

Closing this issue with that becaus of age. Please re-open a new issues if there are still issues.
